---
title: "Simulation Demo"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{simulation_demo}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(cnvseqTV)
```

```{r}
library(readr)
library(magrittr)
```

# CNVseq-TV Demo

In this demo, we make a few deletions in chr1 of the banana genome, and then
generate reads and bams from these, mapped back to the original banana genome.


First, we need to run two commands. First, we use bedtools to generate a list
of genome windows that should be analysed. Second, we run samtools which will
parse the BAM files and generate a count of reads from each window. Obviously,
you'll need both of these things installed, and you need a samtools version of
at least 1.20.

In addition, this is based on a simulation (see ./simulation in the repository
root) written in snakemake, whose outputs aren't included in the repository as
they're too large. If you want to run this, clone the repo and run the
snakemake pipeline yourself.


```{r}
#system("bedtools makewindows -g input/Baxijiao.chr1.fa.fai -w 10000 > tmp/windows.bed")
#system("samtools bedcov -Hjc tmp/windows.bed tmp/bam/*.bam >tmp/window_depths.tsv")
#
# For this demo, we don't actually run that, and instead used a precomputed version at input/window_depths.tsv. If you have the
# required software installed, and have run the simulation workflow, you can of course uncomment the above and actually run it yourself.
```

We then need to instruct cnvseqTV which comparisons to make, and which sample
in each comparison is the control sample. Typically, you'd make a CSV file or
similar in excel, rather than hard-code this into R, but we do it like this to
avoid having another file sitting around and for clarity.

```{r}
comparisons = tibble::tribble(
    ~comparison, ~sample, ~testctrl,
    "deleted_vs_control",  "deleted", "test",
    "deleted_vs_control",  "control", "control",
)
```

Now, we read in the samtools results file, which contains bed-like columns
describing each window, and then columns for each sample with the total
coverage and read counts.

```{r}
rawtbl = read_tsv("input/window_depths.tsv") %>%
    dplyr::glimpse()
```

This is the main function, in which each comparison in the comparisons table is
run on each genome window. The result has one line per window and comparison,
and no filtering happens here, so the result is across the whole genome.

```{r}
wg = calculate_cnvs(rawtbl, comparisons)
```

```{r}
head(wg)
```

```{r}
hit = cnv_rois(wg)
```

```{r}
hit
```

# Plotting

Now, we want to plot our results. To start with, we plot the whole genome. In
our little test, we use only one comparison on chromosome one. In full datasets
this plot will be a matrix of plots, where the multiple samples to be compared
will form rows and the multiple chromosomes will form the columns. Note that in
all cases, a ggplot object is returned, and further customisation and labelling
is possible. Of course, you can also make custom plots of your own, based on
the above two tables (`wg` and `hit` in this notebook).

```{r}
cnv_plot(wg, hit)
```

And now, we zoom in on each of our three regions of interest. Again, these would form

```{r}
cnv_plot(wg, hit, zoom_chrom="chr01", zoom_start=0.5e6, zoom_end=2.5e6) + ggplot2::ylim(-10, 10)
```

```{r}
cnv_plot(wg, hit, zoom_chrom="chr01", zoom_start=3e6, zoom_end=4e6) + ggplot2::ylim(-10, 10)

```

```{r}
cnv_plot(wg, hit, zoom_chrom="chr01", zoom_end=0.5e6) + ggplot2::ylim(-10, 10)
```
